File: C++/Dynamic Programming/120. Triangle.cpp
class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
        if (triangle.size() == 1) return triangle[0][0];
        int res = INT_MAX;
        for (int i = 1; i < triangle.size(); i++) {
            for (int j = 0; j <= i; j++) {
                if (j == 0) triangle[i][j] += triangle[i-1][0];
                else if(j == i) triangle[i][j] += triangle[i-1][i-1];
                else triangle[i][j] += min(triangle[i-1][j-1], triangle[i-1][j]);
                if (i == triangle.size()-1) res = min(res, triangle[i][j]);
            }
        }
        return res;
    }
};

File: Python/Archive/120.triangle.py
# Given a triangle, find the minimum path sum from top to bottom. 
# Each step you may move to adjacent numbers on the row below.

from typing import List

class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        n = len(triangle)
        dp = [0 for i in range(n)]
        dp[0] = triangle[0][0]
        for i in range(1, n):
            row = triangle[i]
            row_len = len(row)
            dp[row_len-1] = row[row_len-1] + dp[row_len-2]
            for j in range(row_len-2,0,-1):
                dp[j] = min(dp[j-1]+row[j], dp[j]+row[j])
            dp[0] += row[0]        
        return min(dp)

solution = Solution()
triangle = [[2],[3,4],[6,5,7],[4,1,8,3]]
print(solution.minimumTotal(triangle))